core/src/gateway/sharded.ts

```diff
@@ -0,0 +1,251 @@
+import { GetGatewayBotPayload } from "../../../types/mod.ts";
+import { EventEmitter } from "../../deps.ts";
+import {
+  GatewayEvents,
+  ShardedGatewayEvents,
+} from "../../types/gateway/events.ts";
+import { RESTClient } from "../rest/rest_client.ts";
+import { Gateway, GatewayOptions } from "./mod.ts";
+
+export interface ShardedGatewayOptions extends Omit<GatewayOptions, "shard"> {
+  shards?: number;
+}
+
+interface Listeners {
+  innerListeners: ((
+    ...args: GatewayEvents[keyof GatewayEvents]
+  ) => void)[];
+  outerListener?: (
+    ...args: ShardedGatewayEvents[keyof ShardedGatewayEvents]
+  ) => void;
+  once: boolean;
+}
+
+export class ShardedGateway extends EventEmitter<ShardedGatewayEvents> {
+  shards: Record<number, Gateway> = {};
+  token: string;
+  intents: number;
+  options: ShardedGatewayOptions;
+  shardCount?: number;
+  #_listeners: Record<string, Listeners[]> = {};
+
+  constructor(
+    token: string,
+    intents: number,
+    options: ShardedGatewayOptions = {},
+  ) {
+    super();
+    this.token = token;
+    this.intents = intents;
+    this.options = options;
+    this.shardCount = options.shards;
+  }
+
+  async getBestShardCount() {
+    const restClient = new RESTClient({ token: this.token });
+    const { shards }: GetGatewayBotPayload = await restClient.get(
+      "/gateway/bot",
+    );
+
+    return shards;
+  }
+
+  async getShardCount() {
+    if (this.shardCount) return this.shardCount;
+    const shardCount = await this.getBestShardCount();
+    this.shardCount = shardCount;
+    return shardCount;
+  }
+
+  async spawn(shardID: number) {
+    const shardCount = await this.getShardCount();
+    this.shards[shardID] = new Gateway(this.token, this.intents, {
+      ...this.options,
+      shard: [shardID, shardCount],
+    });
+  }
+
+  async spawnAll() {
+    const shardCount = await this.getShardCount();
+    for (let i = 0; i < shardCount; i++) {
+      await this.spawn(i);
+    }
+  }
+
+  async destroy(shardID: number) {
+    if (!this.shards[shardID]) return;
+    await this.shards[shardID].disconnect();
+  }
+
+  async destroyAll() {
+    for (const shardID in this.shards) {
```

Comment:
Why not simplify this to avoid the string to number cast
```ts
for (let i = 0; i < this.shard.length; i++) {
  await this.destroy(i);
}
```

Reply:
nice